ibtk/include/ibtk/CartGridPointwiseFunction.h

```diff
@@ -0,0 +1,87 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (c) 2023 - 2023 by the IBAMR developers
+// All rights reserved.
+//
+// This file is part of IBAMR.
+//
+// IBAMR is free software and is distributed under the 3-clause BSD
+// license. The full text of the license can be found in the file
+// COPYRIGHT at the top level directory of IBAMR.
+//
+// ---------------------------------------------------------------------
+
+#ifndef included_IBTK_CartGridPointwiseFunction
+#define included_IBTK_CartGridPointwiseFunction
+
+/////////////////////////////// INCLUDES /////////////////////////////////////
+
+#include <ibtk/CartGridFunction.h>
+#include <ibtk/ibtk_utilities.h>
+
+/////////////////////////////// CLASS DEFINITION /////////////////////////////
+
+namespace IBTK
+{
+/*!
+ * \brief Function signatures for CartGridPointwiseFunction
+ */
+namespace PointwiseFunctions
+{
+using ScalarFcn = std::function<double(double, const VectorNd&, double)>;
+using VectorFcn = std::function<VectorNd(const VectorNd&, const VectorNd&, double)>;
+using OtherFcn = std::function<VectorXd(const VectorXd&, const VectorNd&, double)>;
+using StaggeredFcn = std::function<double(double, const VectorNd&, double, int)>;
```

Comment:
We *could* make this function a Functor and make it a template argument but lets start this off by making it a single `std::function()`. That way we can remove the template on `F` from the class below.

Reply:
Do you mean to make the first argument `double*`? Otherwise this class can't act on a vector (or tensor) as a whole.

For example, there's a few spots in the complex fluid code that computes the square (or exponential or projection) of a tensor on each index. It would be useful to be able to do that without explicitly looping over the hierarchy.

Reply:
No, but perhaps that's possible by some other mechanism: I'll take a look.

Reply:
OK, how about for those we make `VectorXd` etc. a template type of this class? We can assert that the variable has a depth / depth * n_components which matches the function. I'm pretty sure this will still work with a single template version of the function: we just need to internally use length 1 vectors for scalars.

Reply:
So something like this?
```c++
template <typename T>
class CartGridPointwiseFunction : public CartGridFunction
{
public:
    CartGridPointwiseFunction(std::string, std::function<T(const T, const VectorNd&, const int)>);

    void setDataOnPatch() override;
private:
    std::function<T(const T, const VectorNd&,  const int)> d_f;
}
```
I think that would work, but might be confusing for staggered representations. A vector collocated centered quantity would use `CartGridPointwiseFunction<VectorNd>` while a staggered vector would use `CartGridPointwiseFunction<double>`. Then again, I'm not sure the current implementation is better in that regard.